File: src/main/java/io/github/biligoldenwater/midiplayer/utils/particlegui/component/Panel.java
package io.github.biligoldenwater.midiplayer.utils.particlegui.component;

import io.github.biligoldenwater.midiplayer.utils.particlegui.PixelColor;
import io.github.biligoldenwater.midiplayer.utils.particlegui.interfaces.Component;
import org.bukkit.Color;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class Panel extends Component {
    // final option
    private final boolean drawBorder;
    // final data
    private final List<Component> children;
    // option
    private PixelColor borderColor;

    public Panel(int x, int y, int width, int height, boolean drawBorder) {
        super(x, y, width, height, false);
        if (drawBorder) {
            this.clientWidth = Math.max(width - 2, 0);
            this.clientHeight = Math.max(height - 2, 0);
        }

        this.drawBorder = drawBorder;

        this.children = new ArrayList<>();

        this.borderColor = new PixelColor(Color.BLACK);

        this.render();
    }

    public void render() {
        int width = getWidth();
        List<Component> children = new ArrayList<>(this.children);

        this.pixels = new PixelColor[width * getHeight()];

        //region draw children
        for (Component child : children) {
            if (!child.needRender()) continue;

            child.render();
        }
        for (Component child : children) {
            int childX = child.getX() + (drawBorder ? 1 : 0);
            int childY = child.getY() + (drawBorder ? 1 : 0);
            for (int y = 0; y < child.getHeight(); y++) {
                for (int x = 0; x < child.getWidth(); x++) {
                    int posX = childX + x;
                    int posY = childY + y;

                    // 当超出窗口边界时
                    if (posX < (drawBorder ? 1 : 0) || posY < (drawBorder ? 1 : 0)) continue;
                    if (posX >= this.getWidth() - (drawBorder ? 1 : 0) ||
                            posY >= this.getHeight() - (drawBorder ? 1 : 0))
                        continue;

                    PixelColor pixelColor = child.getPixel(x, y);
                    if (pixelColor == null || pixelColor.getAlpha() == 0) continue; // 组件像素颜色为空或完全透明时

                    PixelColor originalPixelColor = getPixel(posX, posY);
                    if (originalPixelColor == null) { // 如果窗口当前像素为空
                        originalPixelColor = pixelColor;
                    } else {
                        originalPixelColor.overlap(pixelColor);
                    }

                    pixels[posX + posY * width] = originalPixelColor;
                }
            }
        }
        //endregion

        //region draw border
        if (drawBorder) {
            for (int y = 0; y < this.getHeight(); y++) {
                pixels[y * width] = borderColor;
                pixels[(this.getWidth() - 1) + y * width] = borderColor;
            }
            for (int x = 0; x < this.getWidth(); x++) {
                pixels[x] = borderColor;
                pixels[x + (this.getHeight() - 1) * width] = borderColor;
            }
        }
        //endregion

        changeApplied();
    }

    @Override
    public boolean needRender() {
        for (Component child : children) {
            if (child.needRender()) return true;
        }
        return super.needRender();
    }

    public void addChild(Component child) {
        onChange();
        child.setParent(this);
        this.children.add(child);
    }

    public void removeChild(Component child) {
        child.setParent(null);
        this.children.remove(child);
    }

    public void forEachChildren(Consumer<? super Component> action) {
        children.forEach(action);
    }

    public int getChildrenSize() {
        return children.size();
    }

    public boolean isDrawBorder() {
        return drawBorder;
    }

    public PixelColor getBorderColor() {
        return borderColor;
    }

    public void setBorderColor(PixelColor borderColor) {
        onChange();
        this.borderColor = borderColor;
    }

    @Override
    public void setSize(int width, int height) {
        super.setSize(width, height);
        if (drawBorder) {
            this.clientWidth = Math.max(width - 2, 0);
            this.clientHeight = Math.max(height - 2, 0);
        }
    }
}

File: src/main/java/io/github/biligoldenwater/midiplayer/utils/particlegui/component/Panel.kt
package io.github.biligoldenwater.midiplayer.utils.particlegui.component

import io.github.biligoldenwater.midiplayer.utils.particlegui.PixelColor
import io.github.biligoldenwater.midiplayer.utils.particlegui.interfaces.Component
import org.bukkit.Color

open class Panel(x: Int, y: Int, width: Int, height: Int, drawBorder: Boolean) : Component(x, y, width, height, false) {
    // final option
    val isDrawBorder: Boolean

    // final data
    val children: MutableList<Component>
    val childrenSize: Int
        get() = children.size

    // option
    var borderColor: PixelColor

    init {
        if (drawBorder) {
            clientWidth = Math.max(width - 2, 0)
            clientHeight = Math.max(height - 2, 0)
        }
        isDrawBorder = drawBorder
        children = ArrayList()
        borderColor = PixelColor(Color.BLACK)
        render()
    }

    override fun render() {
        val width = width
        val children: List<Component> = ArrayList(
            children
        )
        pixels = arrayOfNulls(width * height)

        //region draw children
        for (child in children) {
            if (!child.needRender()) continue
            child.render()
        }
        for (child in children) {
            val childX = child.x + if (isDrawBorder) 1 else 0
            val childY = child.y + if (isDrawBorder) 1 else 0
            for (y in 0 until child.height) {
                for (x in 0 until child.width) {
                    val posX = childX + x
                    val posY = childY + y

                    // 当超出窗口边界时
                    if (posX < (if (isDrawBorder) 1 else 0) || posY < (if (isDrawBorder) 1 else 0)) continue
                    if (posX >= this.width - (if (isDrawBorder) 1 else 0) ||
                        posY >= height - (if (isDrawBorder) 1 else 0)
                    ) continue
                    val pixelColor = child.getPixel(x, y)
                    if (pixelColor == null || pixelColor.alpha == 0.0) continue  // 组件像素颜色为空或完全透明时
                    var originalPixelColor = getPixel(posX, posY)
                    if (originalPixelColor == null) { // 如果窗口当前像素为空
                        originalPixelColor = pixelColor
                    } else {
                        originalPixelColor.overlap(pixelColor)
                    }
                    pixels[posX + posY * width] = originalPixelColor
                }
            }
        }
        //endregion

        //region draw border
        if (isDrawBorder) {
            for (y in 0 until height) {
                pixels[y * width] = borderColor
                pixels[this.width - 1 + y * width] = borderColor
            }
            for (x in 0 until this.width) {
                pixels[x] = borderColor
                pixels[x + (height - 1) * width] = borderColor
            }
        }
        //endregion
        changeApplied()
    }

    override fun needRender(): Boolean {
        for (child in children) {
            if (child.needRender()) return true
        }
        return super.needRender()
    }

    fun addChild(child: Component) {
        onChange()
        child.parent = this
        children.add(child)
    }

    fun removeChild(child: Component) {
        child.parent = null
        children.remove(child)
    }


    override fun setSize(width: Int, height: Int) {
        super.setSize(width, height)
        if (isDrawBorder) {
            clientWidth = Math.max(width - 2, 0)
            clientHeight = Math.max(height - 2, 0)
        }
    }
}